### PR TITLE
Mvfs 3/sonyflake

### PR DESCRIPTION
**Pull request by pihish**
FYI, this actually includes all implementations.  The implementations were close enough where it didn't make sense for me to create a separate branch for each specific one.

One thing I wasn't quite sure about was the static `bootGeneratesId()` method in **GeneratesSnowflakeId**.  Does this method get called on a model's initialization?  I know the `boot()` method will be called on initialization, but didn't find any documentation that said prepending a function with  the keyword _boot_ would result in the same thing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,13 +2,21 @@
...
 
 ## Usage
-
+Include this trait in your Laravel model by adding the following command to the model class:
+```
+use Moves\Snowflake\Traits\GeneratesSnowflakeId;
+```
+Then initialize the generator with the following:
+```
+self::bootGeneratesId(trait_type);
```
Comment (nickrupert7) [inline README.md:18]: The `bootGeneratesId` function is one that is called _automatically_ by Laravel, not manually by the developer.

Let's have a quick call to discuss this later - it shouldn't be a big deal to tweak this and get it working the way it's supposed to.

**Review by nickrupert7 — CHANGES_REQUESTED** [review]
@pihish In general I think the core logic is there, but there are a few things I'd like to see done differently.

Let's schedule a quick call to check in and discuss, rather than posting everything here in GitHub

File: src/Traits/GeneratesSnowflakeId.php [hunk 2]
```diff
@@ -3,28 +3,40 @@
...
+				$generator = $this->_getSnowflakeGenerator($type);
+				$keyField = $model->getKeyName();
+				$model->$keyField = $generator->generate();
+			}
+		});
+
+	}
+
+	// Returns the appropriate generator.  Default generator is the model based generator
+	public function _getSnowflakeGenerator($type): ISnowflakeGenerator{
```
Comment (nickrupert7) [inline src/Traits/GeneratesSnowflakeId.php:25]: In general I'm not a fan of string-keying a setting like this. Instead, an interface which forces the developer to implement the `getSnowflakeGenerator` function with a return type of of `ISnowflakeGenerator`, you are guaranteed to return a valid generator instance. If they pass a string like `twittr` (notice the typo) and then because the `else` statement below defaults to `ModelflakeGenerator`, the actual behavior here would be very confusing, and would be a frustrating bug to track down.

In short, no modifications to this trait should have been necessary

File: src/Generators/ModelflakeGenerator.php [hunk 3]
```diff
@@ -8,17 +8,41 @@
...
+
+	private string $modelName;
+	private int $microtimeLength;
+	private int $identifierLength;
+
+	public function __construct($modelName){
+
+		$this->modelName = $modelName;
+		$this->microtimeLength = 39;
+		$this->identifierLength = 24;
```
Comment (nickrupert7) [inline src/Generators/ModelflakeGenerator.php]: If you're not passing values in through the constructor, they should simply be set as default values where the variables are declared

File: src/Generators/ModelflakeGenerator.php [hunk 4]
```diff
@@ -8,17 +8,41 @@
  * Class ModelflakeGenerator
  *
  * Custom implementation inspired by Twitter Snowflake.
- * Generate Snowflake IDs with a 7-bit prefix identifier for the model that the ID was generated for
```
Comment (nickrupert7) [inline src/Generators/ModelflakeGenerator.php:11]: Why did you remove the "7-bit" description? I left that there to specify how this should be built

File: src/Generators/SonyflakeGenerator.php [hunk 5]
```diff
@@ -11,10 +11,40 @@
...
+
+	private int $microtimeLength;
+	private int $machineLength;
+	private int $sequenceLength;
+
+	public function __construct(){
+
+		$this->microtimeLength = 39;
+		$this->machineLength = 16;
+		$this->sequenceLength = 8;
```
Comment (nickrupert7) [inline src/Generators/SonyflakeGenerator.php]: Same here, default values should not be set in the constructor, but at the variable declaration itself.

File: src/Generators/TwitterSnowflakeGenerator.php [hunk 6]
```diff
@@ -12,10 +12,40 @@
...
+
+	private int $microtimeLength;
+	private int $addressLength;
+	private int $portLength;
+
+	public function __construct(){
+
+		$this->microtimeLength = 41;
+		$this->addressLength = 13;
+		$this->portLength = 9;
```
Comment (nickrupert7) [inline src/Generators/TwitterSnowflakeGenerator.php]: Same here

File: src/Generators/SonyflakeGenerator.php [hunk 7]
```diff
@@ -11,10 +11,40 @@
...
+
+		$microtime = $this->getUnixMicroTimestamp();
+		$remoteAddress = isset($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : 'localhost';
+		$remotePort = isset($_SERVER['REMOTE_PORT']) ? $_SERVER['REMOTE_PORT'] : '80';
+
+		$binMicrotime = str_pad(decbin($microtime), $this->microtimeLength, 0, STR_PAD_LEFT);
+		$binRemoteAddress = str_pad(decbin($remoteAddress), $this->machineLength, 0, STR_PAD_LEFT);
+		$binRemotePort = str_pad(decbin($remotePort), $this->sequenceLength, 0, STR_PAD_LEFT);
+
+		return (int)bindec($binMicrotime . $binRemoteAddress . $binRemotePort);
```
Comment (nickrupert7) [inline src/Generators/SonyflakeGenerator.php:38]: I'm really confused about the implementation of this function... It looks like you are simply concatenating the raw values, some of which are strings, then passing that into `decbin` which expects an integer as its argument. Just from some quick experiments I think this should produce a `TypeError`, and plus I'm not sure that this would produce the correct bit sequencing outlined in the Sonyflake spec.

File: src/Generators/TwitterSnowflakeGenerator.php [hunk 8]
```diff
@@ -12,10 +12,40 @@
...
+
+		$microtime = $this->getUnixMicroTimestamp();
+		$remoteAddress = isset($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : 'localhost';
+		$remotePort = isset($_SERVER['REMOTE_PORT']) ? $_SERVER['REMOTE_PORT'] : '80';
+
+		$binMicrotime = str_pad(decbin($microtime), $this->microtimeLength, 0, STR_PAD_LEFT);
+		$binRemoteAddress = str_pad(decbin($remoteAddress), $this->addressLength, 0, STR_PAD_LEFT);
+		$binRemotePort = str_pad(decbin($remotePort), $this->portLength, 0, STR_PAD_LEFT);
+
+		return (int)bindec($binMicrotime . $binRemoteAddress . $binRemotePort);
```
Comment (nickrupert7) [inline src/Generators/TwitterSnowflakeGenerator.php:39]: Same here

File: src/Generators/ModelflakeGenerator.php [hunk 9]
```diff
@@ -8,17 +8,41 @@
...
+	public function generate(): int{
+
+		$microtime = $this->getUnixMicroTimestamp();
+		$remoteAddress = isset($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : 'localhost';
+		$remotePort = isset($_SERVER['REMOTE_PORT']) ? $_SERVER['REMOTE_PORT'] : '80';
+
+		$binMicrotime = str_pad(decbin($microtime), $this->microtimeLength, 0, STR_PAD_LEFT);
+		$binIdentifier = str_pad(decbin($remoteAddress . $remotePort . $this->modelName), $this->identifierLength, 0, STR_PAD_LEFT);
+
+		return (int)bindec($binMicrotime . $binIdentifier);
```
Comment (nickrupert7) [inline src/Generators/ModelflakeGenerator.php:42]: Same here

**Review by nickrupert7 — CHANGES_REQUESTED** [review]
@pihish I went ahead and added some more detailed comments for posterity and future reference, but I'd still like to have a call to catch up and go over the areas that I think need tweaking
